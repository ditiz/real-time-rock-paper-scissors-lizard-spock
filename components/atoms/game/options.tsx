interface GameOptionsAtomProps {}

const GameOptionsAtom: React.FC<GameOptionsAtomProps> = () => {
  return <></>
}

export default GameOptionsAtomProps
