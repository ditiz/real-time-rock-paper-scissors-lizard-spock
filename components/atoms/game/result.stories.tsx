import { ComponentMeta, ComponentStory } from "@storybook/react"
import { GameResultAtom } from ".."

export default {
  title: "Atoms/Game/Result",
  component: GameResultAtom,
  args: {
    children: "win",
  },
} as ComponentMeta<typeof GameResultAtom>

const Template: ComponentStory<typeof GameResultAtom> = (args) => (
  <GameResultAtom {...args} />
)

export const Default = Template.bind({})
