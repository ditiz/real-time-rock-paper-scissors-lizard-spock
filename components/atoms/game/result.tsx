import { ReactNode } from "react"

interface ResultAtomProps {
  children: ReactNode
}

const ResultAtom: React.FC<ResultAtomProps> = ({ children }) => {
  return <section className="game__result">{children}</section>
}

export default ResultAtom
