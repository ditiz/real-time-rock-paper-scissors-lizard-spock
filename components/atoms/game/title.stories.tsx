import { GameTitleAtom } from ".."

export default {
  title: "Atoms/Game/Title",
  component: GameTitleAtom,
}

export const Default = () => <GameTitleAtom>Title</GameTitleAtom>
