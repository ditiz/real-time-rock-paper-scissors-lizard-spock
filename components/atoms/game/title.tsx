interface GameTitleProps {
  children: React.ReactNode
}

const GameTitle: React.FC<GameTitleProps> = ({ children }) => {
  return <h2 className="game__title">{children}</h2>
}

export default GameTitle
