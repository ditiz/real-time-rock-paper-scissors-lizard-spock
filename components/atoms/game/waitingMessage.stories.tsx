import { ComponentMeta, ComponentStory } from "@storybook/react"
import { WaitingMessageAtom } from ".."

export default {
  title: "Atoms/Game/WaitingMessage",
  component: WaitingMessageAtom,
} as ComponentMeta<typeof WaitingMessageAtom>

const Template: ComponentStory<typeof WaitingMessageAtom> = (args) => (
  <WaitingMessageAtom {...args} />
)

export const Default = Template.bind({})
Default.args = {
  message: "Waiting for another player",
}
