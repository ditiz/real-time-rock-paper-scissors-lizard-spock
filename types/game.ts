export enum Option {
  rock = "rock",
  paper = "paper",
  scissors = "scissors",
  lizard = "lizard",
  spock = "spock",
}

export enum GameResult {
  tie = "tie",
  win = "win",
  lose = "lose",
}
