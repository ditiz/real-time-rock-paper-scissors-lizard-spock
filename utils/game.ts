import { GameResult, ImageOption, Option } from "@/types/game"

const rock = {
  name: "rock",
  path: "icons/rock.svg",
  alt: "rock icon",
  value: Option.rock,
}

const paper = {
  name: "paper",
  path: "icons/paper.svg",
  alt: "paper icon",
  value: Option.paper,
}

const scissors = {
  name: "scissors",
  path: "icons/scissors.svg",
  alt: "scissors icon",
  value: Option.scissors,
}

const lizard = {
  name: "lizard",
  path: "icons/lizard.svg",
  alt: "lizard icon",
  value: Option.lizard,
}

const spock = {
  name: "spock",
  path: "icons/spock.svg",
  alt: "spock icon",
  value: Option.spock,
}

export const GameAdvanceOptions: ImageOption[] = [
  rock,
  paper,
  scissors,
  lizard,
  spock,
]

export const GameClassicOptions: ImageOption[] = [rock, paper, scissors]

// Object that define against what an option win
const actions = {
  [Option.rock]: [Option.scissors, Option.lizard],
  [Option.paper]: [Option.rock, Option.spock],
  [Option.scissors]: [Option.paper, Option.lizard],
  [Option.lizard]: [Option.paper, Option.spock],
  [Option.spock]: [Option.rock, Option.scissors],
}

/**
 * Say who win the game
 * @param choice
 * @param opponentChoice
 * @returns 0 for tie, 1 for win, -1 for lose
 */
export function playGame(choice: Option, opponentChoice: Option): GameResult {
  // Check tie
  if (choice === opponentChoice) {
    return GameResult.tie
  }

  // Check if win
  if (actions[choice].includes(opponentChoice)) {
    return GameResult.win
  }

  // If no tie or win, it's lose
  return GameResult.lose
}
