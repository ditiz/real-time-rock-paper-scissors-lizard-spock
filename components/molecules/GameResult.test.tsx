import { GameResult } from "@/types/game"
import "@testing-library/jest-dom"
import { render, screen } from "@testing-library/react"
import GameResultMolecule from "./GameResult"

it("renders DisplayGameResult unchanged", () => {
  const { container } = render(<GameResultMolecule result={GameResult.win} />)
  expect(container).toMatchSnapshot()
})

test("win", async () => {
  render(<GameResultMolecule result={GameResult.win} />)
  expect(screen.getByText("You win !")).toBeTruthy()
})

test("tie", async () => {
  render(<GameResultMolecule result={GameResult.tie} />)
  expect(screen.getByText("It's a tie")).toBeTruthy()
})

test("lose", async () => {
  render(<GameResultMolecule result={GameResult.lose} />)
  expect(screen.getByText("You lose...")).toBeTruthy()
})
