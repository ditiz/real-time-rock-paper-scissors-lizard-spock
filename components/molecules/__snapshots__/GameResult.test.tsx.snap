// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DisplayGameResult unchanged 1`] = `
<div>
  <article
    class="game__result"
  >
    <section>
      You win !
    </section>
  </article>
</div>
`;
